src/farming/library.cairo

```diff
@@ -0,0 +1,1039 @@
+        );
+    }
+
+    //
+    // Externals
+    //
+
+    func set_prices{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        times_len: felt, times: felt*, prices_len, prices: felt*
+    ) {
+        alloc_locals;
+
+        Array.store(key=TIME_SK, len=times_len, values=times);
+        Array.store(key=PRICE_SK, len=prices_len, values=prices);
+
+        return ();
+    }
+
+    func add_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        time: felt, price: felt
+    ) {
+        alloc_locals;
+
+        // [Check] Time is later than the project start time
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (first_time) = ICarbonableProject.getStartTime(
+            contract_address=contract_address, slot=slot
+        );
+        with_attr error_message("CarbonableFarming: time is sooner than the project start time") {
+            assert_le(first_time + 1, time);
+        }
+
+        // [Check] First time to store
+        let (len) = Array.read_len(key=TIME_SK);
+        let (current_time) = get_block_timestamp();
+        if (len == 0) {
+            // [Effect] Update storage
+            Array.add(key=TIME_SK, value=first_time);
+            Array.add(key=TIME_SK, value=time);
+            Array.add(key=PRICE_SK, value=price);
+            Array.add(key=PRICE_SK, value=price);
+
+            // [Event] Emit event
+            PriceUpdate.emit(time=time, price=price);
+
+            return ();
+        }
+
+        // [Check] Time is later than the last stored time
+        let (last_time) = Array.read(key=TIME_SK, index=len - 1);
+        with_attr error_message("CarbonableFarming: time is later than the last stored time") {
+            assert_le(last_time + 1, time);
+        }
+
+        // [Check] Absorption is captured between these times
+        let (initial_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=last_time
+        );
+        let (final_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=time
+        );
+        with_attr error_message("CarbonableFarming: the period must capture absorption") {
+            assert_le(initial_absorption + 1, final_absorption);
+        }
+
+        // [Effect] Update storage
+        Array.add(key=TIME_SK, value=time);
+        Array.add(key=PRICE_SK, value=price);
+
+        // [Event] Emit event
+        PriceUpdate.emit(time=time, price=price);
+
+        return ();
+    }
+
+    func update_last_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        time: felt, price: felt
+    ) {
+        alloc_locals;
+
+        // [Check] At least 2 prices stored
+        let (len) = Array.read_len(key=TIME_SK);
```

Comment:
If I follow the comment, the check should be key=PRICE_SK instead of key=TIME_SK ?

Reply:
Same have above.